feat(Peer Chat): Send a notification when a Peer Chat message is sent

geoffreykwan opened the pull request:
## Changes
- When a Peer Chat message is sent, a notification is also sent to Peer Chat members

## Test
- Make sure a notification shows up when you are not on the Peer Chat step
- Make sure a notification does not show up when you are on the Peer Chat step
- Make sure only one notification shows up even if multiple messages are sent
- This scenario
  - Student is on a Peer Chat step
  - Student receives a notification for the Peer Chat step but we don't show it to them because they are on the same step as the notification
  - Refresh the VLE
  - The notification should still not show up

Closes #509

geoffreykwan committed:
feat(Peer Chat): Send a notification when a Peer Chat message is sent

#509

geoffreykwan committed:
fix(Notifications): Dismiss new notification if student is on same step

#509

hirokiterashima commented:
In this scenario:
- There are three students (A,B,C) in a chat together
- A and B are on the chat step, C is on a different step
- A sends a message to chat
- B sends a message to chat

C only receives one message "A sent a chat message". This is not wrong, but it would be nice if the notification reflects the fact that B also sent a chat message.

In this case, should C see 
- one notification "A and B sent a chat message"? 
- two notifications "A sent a chat message", "B sent a chat message"
- keep as is?

geoffreykwan committed:
Merge branch 'issue-370-create-peer-chat-component' into issue-509-peer-chat-message-notification

hirokiterashima committed:
Merge branch 'issue-370-create-peer-chat-component' into issue-509-peer-chat-message-notification

hirokiterashima reviewed (changes requested):
When a teacher is a participant of a chat, they will receive a new notification. Can you change it so that clicking on it will take the teacher to the step where the chat is on? Right now it just dismisses the notification.

geoffreykwan committed:
feat(Notification): Make clicking notification send teacher to step

When a teacher clicks on a notification, it will send them to the grade
by step view for the step the notification is associated with.

#509

geoffreykwan committed:
feat(Peer Chat): Only send notification to teacher if teacher posted

Previously we were sending a notification to the teacher even if the
teacher never sent a message to the group the student is in.

#509

geoffreykwan committed:
fix(Notification): Prevent showing multiple from the same source

If the teacher receives multiple notifications from the same source, we
will only show one to them. If the teacher dismisses a notification, we
also dismiss all other notifications from the same source.

#509

hirokiterashima committed:
Fix typos, added private signature and types, refactored code.
Remove unused file.

hirokiterashima committed:
Merge branch 'issue-370-create-peer-chat-component' into issue-509-peer-chat-message-notification

hirokiterashima committed:
Fix issue where new notification was not being dismissed
Refactor getActiveNotificationsWithSameSource()

hirokiterashima commented:
I noticed these things on the teacher’s end:
- they always receive a notification from a chatroom they were in even if they’re on the chat step
- when they submit a new message to the chat, it does not create a notification for the students

These can be handled in a separate issue.
